zapcore/sampler.go

```diff
@@ -81,33 +81,96 @@ func (c *counter) IncCheckReset(t time.Time, tick time.Duration) uint64 {
 	return 1
 }
 
-type sampler struct {
-	Core
+// SamplingDecision is a decision made by sampler.
+type SamplingDecision uint8
 
-	counts            *counters
-	tick              time.Duration
-	first, thereafter uint64
+const (
+	// LogDropped indicates that the Sampler dropped a log entry.
+	LogDropped SamplingDecision = iota
+	// LogSampled indicates that the Sampler sampled a log entry.
+	LogSampled
```

Comment:
something to discuss
- it would be useful to know when a log was first dropped, and then first log after the sample again. e.g., if you want to add a special log that said "message X is now being dropped, no further logs"?
- doing this in the hook requires doing the same thing the sampler does, probably more inefficiently (a map by message)

Reply:
Just an idea: If we make SamplingDecision a bitfield, we could do that separately and re-use sampler logic

```
LogDropped SamplingDecision = 1 << iota
LogSampled
LogDropInitial
LogDropThereafter
```